addons/knowledge/static/src/webclient/commands/knowledge_providers.js

```diff
@@ -0,0 +1,148 @@
+/** @odoo-module */
+
+
+import { _lt } from "@web/core/l10n/translation";
+import { registry } from "@web/core/registry";
+import { HotkeyCommandItem } from "@web/core/commands/default_providers";
+
+const { Component } = owl;
+const { xml } = owl.tags;
+
+
+class KnowledgeCommand extends Component {}
+KnowledgeCommand.template = xml`
+    <div class="o_command_left">
+        <icon t-att-class="'pr-2 fa ' + props.icon_string"/>
+        <span t-esc="props.name" class="pr-2"/>
+        <icon t-if="props.isFavourite" class="fa fa-star o_favorite pr-2"/>
+        <span t-if="props.parentName" t-esc="'— '" class="text-muted small pr-2" />
+        <span t-if="props.parentName" t-esc="props.parentName" class="text-muted small" />
+    </div>
+`;
+
+class Knowledge404Command extends Component {}
+Knowledge404Command.template = xml`
+    <div class="o_command_hotkey">
+        <span>
+            No Article found. <span class="text-primary">Create "<u t-esc="props.articleName"/>"</span>
+        </span>
+    </div>
+`;
+
+class KnowledgeExtraCommand extends HotkeyCommandItem {}
+KnowledgeExtraCommand.template = xml`
+    <div class="o_command_hotkey">
+        <span>
+            <icon class="fa fa-arrows-alt pr-2" />
+            <t t-esc="props.name" />
+        </span>
+        <span>
+            <t t-foreach="getKeysToPress(props)" t-as="key">
+                <kbd t-esc="key" />
+                <span t-if="!key_last"> + </span>
+            </t>
+        </span>
+    </div>
+`;
+const commandSetupRegistry = registry.category("command_setup");
+commandSetupRegistry.add("?", {
+    emptyMessage: _lt("No article found."),
+    name: _lt("articles"),
+});
+
+const commandProviderRegistry = registry.category("command_provider");
+commandProviderRegistry.add("knowledge", {
```

Comment:
+ add comments on what you do in those two functions.. 
While we are at it, we could put all the code in separated functions and call 
`commandProviderRegistry.add("knowledge", nameOfTheFunction)`

Reply:
the second argument of `.add(` is not a function but a dictionary, so should I define a dictionary outside and call it there?
What exactly would be the advantage? I don't see how it really improves code readability and it won't be re-used anywhere else. 

In the process of writing comments :)!